pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java

```diff
@@ -0,0 +1,328 @@
+
+import com.google.common.base.Preconditions;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.core.common.DataSource;
+import org.apache.pinot.core.operator.blocks.ProjectionBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.utils.ByteArray;
+
+
+/**
+ * <code>BinaryOperatorTransformFunction</code> abstracts common functions for binary operators (=, !=, >=, >, <=, <)
+ * The results are in boolean format and stored as an integer array with 1 represents true and 0 represents false.
+ */
+public abstract class BinaryOperatorTransformFunction extends BaseTransformFunction {
+
+  protected TransformFunction _leftTransformFunction;
+  protected TransformFunction _rightTransformFunction;
+  protected FieldSpec.DataType _leftDataType;
+  protected FieldSpec.DataType _rightDataType;
+  protected int[] _results;
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    // Check that there are exact 2 arguments
+    Preconditions
+        .checkArgument(arguments.size() == 2, "Exact 2 arguments are required for binary operator transform function");
+    _leftTransformFunction = arguments.get(0);
+    _rightTransformFunction = arguments.get(1);
+    _leftDataType = _leftTransformFunction.getResultMetadata().getDataType();
+    _rightDataType = _rightTransformFunction.getResultMetadata().getDataType();
+    // Data type check: left and right types should be compatible.
+    if (_leftDataType == FieldSpec.DataType.BYTES || _rightDataType == FieldSpec.DataType.BYTES) {
+      Preconditions.checkState(_leftDataType == FieldSpec.DataType.BYTES && _rightDataType == FieldSpec.DataType.BYTES,
+          String.format(
+              "Unsupported data type for comparison: [Left Transform Function [%s] result type is [%s], Right Transform Function [%s] result type is [%s]]",
+              _leftTransformFunction.getName(), _leftDataType, _rightTransformFunction.getName(), _rightDataType));
+    }
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return INT_SV_NO_DICTIONARY_METADATA;
+  }
+
+  protected void fillResultArray(ProjectionBlock projectionBlock) {
+    if (_results == null) {
+      _results = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    int length = projectionBlock.getNumDocs();
+    switch (_leftDataType) {
+      case INT:
+        int[] leftIntValues = _leftTransformFunction.transformToIntValuesSV(projectionBlock);
+        switch (_rightDataType) {
+          case INT:
+            int[] rightIntValues = _rightTransformFunction.transformToIntValuesSV(projectionBlock);
+            for (int i = 0; i < length; i++) {
+              _results[i] = getBinaryFuncResult(Integer.compare(leftIntValues[i], rightIntValues[i]));
```

Comment:
Generally we should do this what databases actually do -- generate code at runtime for the exact types to avoid the interpretation overhead of gigantic switch blocks.
Since we don't have that yet, may be we can consider (not necessarily in this PR), hand-writing type specific versions of this  -- IntBinaryOperatorTransformFunction, LongBinaryOperatorTransformFunction so that during execution there are minimal (probably none) branches and there is no interpretation overhead at the expense of duplicate code.

Reply:
I feel those kind of nested function calls would be better to be handle by some middle layer to handle the type conversion. E.g `_rightTransformFunction.transformToIntValuesSV(projectionBlock);`
could become:
`BlockProvider.get(_rightTransformFunction.transformToSV(projectionBlock), FieldType.INT);`
This middle layer will handle fetch from the underlying block and do type conversion to my expected results format.

Reply:
Sounds good, I will double check with that logic